src/sql/src/query_model/hir.rs

```diff
@@ -431,3 +452,137 @@ impl FromHir {
         }
     }
 }
+
+/// Special transformations to fix semantically invalid constructions allowed
+/// temporarily for the sake of keeping the generator above as simple as possible.
+mod fixup {
+
+    use super::*;
+
+    pub(crate) fn fixup_model(model: &mut Model) {
+        let mut windowing_fixup = Windowing::new();
+
+        let boxes = model.boxes.keys().cloned().collect_vec();
+        for box_id in boxes {
+            if windowing_fixup.condition(model, box_id) {
+                windowing_fixup.action(model, box_id);
+            }
+        }
+    }
+
+    struct Windowing {
+        /// Window function expression found in a select box that
+        /// must be pushed down to a Windowing box
+        to_pushdown: BTreeSet<BoxScalarExpr>,
+    }
+
+    impl Windowing {
+        fn new() -> Self {
+            Self {
+                to_pushdown: BTreeSet::new(),
+            }
+        }
+
+        fn condition(&mut self, model: &Model, box_id: BoxId) -> bool {
+            self.to_pushdown.clear();
+            let b = model.get_box(box_id);
+            if b.is_select() {
+                let _ = b.visit_expressions(&mut |expr: &BoxScalarExpr| -> Result<(), ()> {
+                    expr.visit(&mut |expr: &BoxScalarExpr| {
+                        if let BoxScalarExpr::Windowing(_) = expr {
+                            self.to_pushdown.insert(expr.clone());
+                        }
+                    });
+                    Ok(())
+                });
+
+                !self.to_pushdown.is_empty()
+            } else {
+                false
+            }
+        }
+
+        fn action(&mut self, model: &mut Model, box_id: BoxId) {
```

Comment:
Please add a summary of what this does.

What I believe this action does: 
Adds two boxes:
1) a select box that projects all the columns of the input quantifiers of the original box.
2) a window box that projects all the columns of (1) + all window expressions.
And changes the original box so that its only quantifier is a Foreach quantifier pointing to (2).

Reply:
Yeah, this entire module really deserves better docs.